core(stack-packs): move to stack-packs npm module

wardpeet opened the pull request:
**Summary**
Move to stack-packs as an NPM module to make integration/updating these packs easier.  I update the collect-strings script to look into node_modules instead of the stack-packs folder.

All files have been converted to the new syntax, and everything seems to be working as expected. 

**Related Issues/PRs**
[stack-packs PR]( https://github.com/GoogleChrome/lighthouse-stack-packs/pull/38 )

wardpeet commented on lighthouse-core/lib/i18n/locales/ar-XB.json:
```diff
@@ -1460,139 +1460,139 @@
     "message": "‏‮Warnings‬‏: "
   },
-  "stack-packs/packs/amp.js | efficient_animated_content": {
+  "node_modules/lighthouse-stack-packs/packs/amp.js | efficient_animated_content": {
```
All of these keys have changed as the directory of the strings have changed. Is this breaking? Or shouldn't I care about this?

connorjclark replied:
I do prefer as you have it now, but @exterkamp will know if this will recreate work for our translators.

exterkamp replied:
No, these keys only matter to us :smile: Should be a no-op for translators.

wardpeet commented on types/lighthouse-stack-packs/index.d.ts:
```diff
@@ -0,0 +1,30 @@
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
+ */
+
+declare module 'lighthouse-stack-packs/packs/amp' {
```
I didn't want to add these into the `lighthouse-stack-packs` as it adds extra complexity for contributors.

I do think we might want to create a script that generates these definitions. Or at least a script that checks if this file contains all pack definitions.

connorjclark replied:
We don't need a script for this. It's super easy to do manually and we don't update this often.

I'm not sure I agree with keeping this in `lighthouse` to avoid complexity. Most contributions won't be _adding_ frameworks. And if one does, then it's rather straightforward. There are other more complex things that must be done to add a framework, such as changes to `js-library-detector`.

wardpeet commented on stack-packs/packs/angular.js:
```diff
@@ -7,30 +7,14 @@
+  iconDataURL: icon,
   title: 'Angular',
   descriptions: {
     'total-byte-weight': str_(UIStrings.total_byte_weight),
```
Having all these definitions added here manually might lead to missed values when the NPM package is updated. This problem might occur when a new audit description is added between versions. 

Technically we could loop over all keys of the `UI_Strings` export of the stack-pack. Sadly we do have some outliers like `UIStrings.unminified_warning`, where this string is used for two audits. I do think we keep them as separate strings inside the stack pack. We could dedupe the same strings before sending them to the translators.

exterkamp replied:
Could this be populated automatically? Have a map <stack-pack-string-name, audit-id> in the stack pack repo and build the list here dynamically? `'<audit-id>': str_(UIStrings.<stack-pack-string-name>),`

wardpeet replied:
Yes we can, I can put it all in the stack-packs repo, wasn't sure we wanted to.

connorjclark commented on lighthouse-core/scripts/i18n/collect-strings.js:
```diff
@@ -561,7 +563,8 @@ if (require.main === module) {
 
-  const stackPackStrings = collectAllStringsInDir(path.join(LH_ROOT, 'stack-packs/packs'));
+  // const stackPackStrings = collectAllStringsInDir(path.join(LH_ROOT, 'stack-packs/packs'));
+  const stackPackStrings = collectAllStringsInDir(path.join(path.dirname(require.resolve('lighthouse-stack-packs')), 'packs'));
```
this is more readable as 

```
`${LH_ROOT}/node_modules/lighthouse-stack-packs/packs`
```

wardpeet commented:
we need to publish the stackpacks npm package first :) So the PR in stack-packs repo needs to be merged and published by the team that's going to own it. Maybe @housseindjirdeh

paulirish commented on lighthouse-core/scripts/i18n/collect-strings.js:
```diff
@@ -565,7 +567,8 @@ if (require.main === module) {
   console.log('Collected from LH core!');
 
-  const stackPackStrings = collectAllStringsInDir(path.join(LH_ROOT, 'stack-packs/packs'));
+  const stackPacksRoot = `${LH_ROOT}/node_modules/lighthouse-stack-packs/packs`;
```
node modules paths are weird so I prefer using something like 

```suggestion
  const stackPacksRoot = `${require.resolve('lighthouse-stack-packs')}/packs`;
```

wardpeet replied:
haha @connorjclark said I had to change it into LH_ROOT :joy:

paulirish replied:
hahah ah i see that now. :)

sorry connor but we've seen node_module paths be unreliable to guess. so best to rely on `require.resolve()`

connorjclark replied:
good point, for ex: if you yarn link this pkg

wardpeet commented:
I moved everything to the stack-packs npm package and deleted the local package. The npm package now contains icon, id, title & UIstrings. The uistrings are translated in ` lighthouse-core/lib/stack-packs.js`

connorjclark commented on lighthouse-core/lib/stack-packs.js:
```diff
@@ -60,11 +61,29 @@ function getStackPacks(pageStacks) {
       continue;
     }
 
+    // create i18n handler to get translated strings
```
nit: can you update all these comments to be like:

```
// Start with capital letter. End with punctuation.
```

connorjclark commented on lighthouse-core/scripts/i18n/collect-strings.js:
```diff
@@ -565,7 +568,8 @@ if (require.main === module) {
   console.log('Collected from LH core!');
 
-  const stackPackStrings = collectAllStringsInDir(path.join(LH_ROOT, 'stack-packs/packs'));
+  const stackPacksRoot = path.join(path.dirname(require.resolve('lighthouse-stack-packs')), 'packs');
```
`path.join(a, b)` -> `${a}/${b}`. @brendankenny taught me that one. path delimiters are normalized by node so no issue for windows. this is given as input to glob so I _think_ it should still work ... if not i'll eat my words :)

connorjclark replied:
hahahha https://github.com/GoogleChrome/lighthouse/pull/10138#discussion_r371489799 wait nvm ignore me

connorjclark commented:
looking good. let's go get that stack pack PR merged, we ready for that?

wardpeet commented:
@connorjclark I've published the npm package locally and tested it on LH and it works. If we're ok with the code & structure we can merge :)

wardpeet commented:
I've published a canary build of lighthouse-stack-packs. You can view it here https://www.runpkg.com/?lighthouse-stack-packs@0.0.1-0/index.js.

I also updated this branch so everything should work

connorjclark commented:
@wardpeet `yarn update:sample-json`

wardpeet commented:
I do not really know why these tests are failing 😬 As nothing changed on that part.

connorjclark commented:
Some of those errors seem intermittent. I found one that looks real:

```
Error: require.resolve is not a function
```

We can't use `require.resolve` in the browser. Maybe try this:

```

const str_ = i18n.createMessageInstanceIdFn(
      `node_modules/lighthouse-stack-packs/packs/${matchedPack.id}`,
      matchedPack.UIStrings
    );
```

wardpeet commented on lighthouse-core/lib/stack-packs.js:
```diff
@@ -5,8 +5,21 @@
+const stackPacks = require('lighthouse-stack-packs');
+const i18n = require('./i18n/i18n.js');
+
+/**
+ * Resolve a module on web and node
+ * @param {string} module
+ */
+function resolve(module) {
+  if (!require.resolve) {
+    return `node_modules/${module}`;
+  }
+
+  return require.resolve(module);
+}
```
Added a require.resolve to make sure we still do require.resolve in node land but we don't in web land.

connorjclark commented:
> But, certainly we can't use require.resolve in the browser, so this wouldn't work in CDT.

hmmm, actually I think this was wrong. `require.resolve` actually will still exist in the bundle–we rely on this in the config code. We just need to enumerate each file in `lighthouse-stack-packs/packs/*` in `build-bundle.js`, much like we do for the gatherers and audits (`bundle.require`). We don't have a test for the CDT integration to make this obvious, but it will be needed for the roll to CDT to work.

Sorry for back and forth. I think you can just use `require.resolve`. Let's figure out why that dbw test was failing ... can you merge master back to this branch? We got some new niceties like skipping the proto round trip and using GH actions.

connorjclark commented:
continuing in #11370
